accounts/scwallet/wallet.go

```diff
@@ -0,0 +1,1071 @@
+	deriveReq      chan chan struct{}        // Channel to request a self-derivation on
+	deriveQuit     chan chan error           // Channel to terminate the self-deriver with
+}
+
+// NewWallet constructs and returns a new Wallet instance.
+func NewWallet(hub *Hub, card *pcsc.Card) *Wallet {
+	wallet := &Wallet{
+		Hub:  hub,
+		card: card,
+	}
+	return wallet
+}
+
+// transmit sends an APDU to the smartcard and receives and decodes the response.
+// It automatically handles requests by the card to fetch the return data separately,
+// and returns an error if the response status code is not success.
+func transmit(card *pcsc.Card, command *commandAPDU) (*responseAPDU, error) {
+	data, err := command.serialize()
+	if err != nil {
+		return nil, err
+	}
+
+	responseData, _, err := card.Transmit(data)
+	if err != nil {
+		return nil, err
+	}
+
+	response := new(responseAPDU)
+	if err = response.deserialize(responseData); err != nil {
+		return nil, err
+	}
+
+	// Are we being asked to fetch the response separately?
+	if response.Sw1 == sw1GetResponse && (command.Cla != claISO7816 || command.Ins != insGetResponse) {
+		return transmit(card, &commandAPDU{
+			Cla:  claISO7816,
+			Ins:  insGetResponse,
+			P1:   0,
+			P2:   0,
+			Data: nil,
+			Le:   response.Sw2,
+		})
+	}
+
+	if response.Sw1 != sw1Ok {
+		return nil, fmt.Errorf("Unexpected insecure response status Cla=0x%x, Ins=0x%x, Sw=0x%x%x", command.Cla, command.Ins, response.Sw1, response.Sw2)
+	}
+
+	return response, nil
+}
+
+// applicationInfo encodes information about the smartcard application - its
+// instance UID and public key.
+type applicationInfo struct {
+	InstanceUID []byte `asn1:"tag:15"`
+	PublicKey   []byte `asn1:"tag:0"`
+}
+
+// connect connects to the wallet application and establishes a secure channel with it.
+// must be called before any other interaction with the wallet.
+func (w *Wallet) connect() error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	appinfo, err := w.doselect()
+	if err != nil {
+		return err
+	}
+
+	channel, err := NewSecureChannelSession(w.card, appinfo.PublicKey)
+	if err != nil {
+		return err
+	}
+
+	w.PublicKey = appinfo.PublicKey
+	w.log = log.New("url", w.URL())
+	w.session = &Session{
+		Wallet:  w,
+		Channel: channel,
+	}
+	return nil
+}
+
+// doselect is an internal (unlocked) function to send a SELECT APDU to the card.
+func (w *Wallet) doselect() (*applicationInfo, error) {
+	response, err := transmit(w.card, &commandAPDU{
+		Cla:  claISO7816,
+		Ins:  insSelect,
+		P1:   4,
+		P2:   0,
+		Data: appletAID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	appinfo := new(applicationInfo)
+	if _, err := asn1.UnmarshalWithParams(response.Data, appinfo, "tag:4"); err != nil {
+		return nil, err
+	}
+	return appinfo, nil
+}
+
+// ping checks the card's status and returns an error if unsuccessful.
+func (w *Wallet) ping() error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	// We can't ping if not paired
+	if !w.session.paired() {
+		return nil
```

Comment:
Shouldn't that be an error return?

Reply:
nope, because this is called from a polling function that checks the status of the wallet to be reported in `personal`. Returning an error here would mean periodically displaying "not paired" on the console. For all we know, the user has only inserted an unpaired card in the reader, but maybe it is intended to be used by another software that is running alongside geth. So the right approach here is to not report any error.

If the user then tries to use the card by calling `openWallet`, then we'll know that the lack of pairing is an issue, and will report the error then.